### ISSUE TITLE
Questions about Loss_dur

### ISSUE THREAD

**Issue by Liujingxiu23** [jaywalnut310/vits#13]
As in paper Loss_dur  is the negative variational lower bound of equation 7 
<img width="272" alt="equation-7" src="https://user-images.githubusercontent.com/49260794/123195405-b5069a80-d4da-11eb-9090-2512c43d1492.PNG">

Loss_dur = -log(p/q)=logq - logp

Is the logq(x) computed as follows:
<img width="256" alt="equation-8" src="https://user-images.githubusercontent.com/49260794/123195481-d2d3ff80-d4da-11eb-9554-fe972f18de8d.PNG">

In the code:
<img width="774" alt="equation-9" src="https://user-images.githubusercontent.com/49260794/123195826-67d6f880-d4db-11eb-80af-cdfb5ba187df.PNG">

Why in the computation of logq, " **-** logdet_tot_q" not  " **+** logdet_tot_q" ?

**Comment by nartes** [comment]
@Liujingxiu23 did you manage to train the model from scratch? what is the value for dur loss in your logs? i get it as a negative value pretty soon. synthesis does have duration prediction issues indeed.

**Comment by Liujingxiu23** [comment]
@nartes Yes, I can train from scratch successfully, and the synthesized wavs are good. Though I do not understand the loss computation, the synthesized wavs shows that the duration model is right.

You can take the duration loss picture in https://github.com/jaywalnut310/vits/issues/14 as reference.  When converging， the duration loss is about 1.1

**Comment by nartes** [comment]
@Liujingxiu23 did you train ljs or vctk setup? i'm trying vctk one at the moment. I've turned fp16_run off (did you use it in your case), what batch size did you use?

**Comment by nartes** [comment]
![image](https://user-images.githubusercontent.com/902947/130436539-10688a69-aa11-4a17-8786-5dc82fd11332.png)
Yeah, something like from 4k steps it gets negative.

**Comment by Liujingxiu23** [comment]
@nartes  I did not train LJ or VCTK, I trained on Chinese dataset.  I only revised code related to language, and used the default model config.  "fp16_run": true," as default.

**Comment by nartes** [comment]
Okay, segment_size and batch_size? What GPU VRAM size you use? I have 8GiB and by default this model is too big. Also, get some run time errors from pytorch with fp16, that's why turned it off.

**Comment by nartes** [comment]
Tried another config, 4k steps, dur is not negative. will train for 50K to see further.
![image](https://user-images.githubusercontent.com/902947/130436928-b954d740-4657-4201-825a-af54f0a5261a.png)
This is a spectrogram of both models says "good evening", scratch.ogg from current model sounds a littlbe bit like that. Will see the result in 8 hours from now.

**Comment by nartes** [comment]
![image](https://user-images.githubusercontent.com/902947/130437082-48bb666a-c578-4550-a621-9d3dc592c1fa.png)
It's a current tensorboard.

**Comment by Liujingxiu23** [comment]
I changed only the sampling rate to 16k, and left all other parameters as default. 
"sampling_rate": 16000,
"filter_length": 1024,
 "hop_length": 256,
"win_length": 1024,
"n_mel_channels": 80,

The training of vits really consumes resources and a little time comsuming. I use 2 GPU of V100(32G). And the batch size of 64. 

The another mentioned in other issues that the synthesized wavs can achieve good performance at 180k steps (G_180000.pth). My testes on Chinese matches this statement。
But at the very beginning of the training ,for example, G_10000.pth, you can synthesized the test wavs to see whether the model is training normally. At G_10000.pth, though the synthesized wavs are not very good,  they are totally understandable, and the duration is right.

**Comment by Liujingxiu23** [comment]
@nartes  I guess  you may take model config and training config as default and check if anything wrong at data preprocess .
My duration loss:
<img width="203" alt="duration" src="https://user-images.githubusercontent.com/49260794/130438893-87e50061-55fb-44dd-98f7-b0e1e0946f07.PNG">

**Comment by nartes** [comment]
@Liujingxiu23 well, just checked a current result.
![image](https://user-images.githubusercontent.com/902947/130492922-5cca9896-bff3-43ff-a9d2-701b6e646f89.png)

still dur loss gets negative at some points. tried to listen to the audio. complete gibberish.
![image](https://user-images.githubusercontent.com/902947/130492979-85a87252-6a5e-4a3b-b446-56cc972ff4f9.png)

I don't have 2x32GiB GPU, only 8GiB one. So will try to enable fp16, maybe it is important. Also, think about switching to 8KHz sampling rate, at the stage, should allow bigger batch size. By default, model runs only with batch_size==1 for me. Which makes no sense for SGD, cause lots of variance.

**Comment by nartes** [comment]
Started 8KHz, ljs_nosdp, batch_size==1, fp16 == false
![image](https://user-images.githubusercontent.com/902947/130503290-2b6afa63-d148-4fdf-8693-2bfa06cd03d0.png)

![image](https://user-images.githubusercontent.com/902947/130503370-815318f3-db2b-4c70-9b8d-fb429dea77c1.png)
Will check the result in 8hours from now.

**Comment by nartes** [comment]
![image](https://user-images.githubusercontent.com/902947/130556020-4debfd70-f21a-46da-819e-446de11a0f1e.png)

![image](https://user-images.githubusercontent.com/902947/130556046-98e9f8aa-371f-458b-abbd-d7ad9c3aa3e2.png)

![image](https://user-images.githubusercontent.com/902947/130556085-ee70230b-2d39-4c54-b7ac-bc5ff124288f.png)

Looks a bit better, yet generated sound is non sense. Maybe need to fine tune fft parameters as well, cause it can be too rough for 8192 sample rate.

Will stop for now on this model. Looks like a waste of time.

**Comment by Liujingxiu23** [comment]
I guess batch_size=1 may raises some problem. In consideration of the limit of the GPU resources, maybe you can change the dataloader code. The current dataloader load the whole length of the wav, but in training only "segment" is used. You may change to sample segment in dataloader and only push "segment". I tried this way in other E2E TTS training. In one GPU(16G), batch_size=64 can be used.

Why set sampel-rate=8192, 8192 is your target sampel-rate ? 

What is your “ "upsample_initial_channel"？Change it to 256 / 128 may get the model smaller?

**Comment by p0p4k** [comment]
> As in paper Loss_dur is the negative variational lower bound of equation 7 <img alt="equation-7" width="272" src="https://user-images.githubusercontent.com/49260794/123195405-b5069a80-d4da-11eb-9090-2512c43d1492.PNG">
> 
> Loss_dur = -log(p/q)=logq - logp
> 
> Is the logq(x) computed as follows: <img alt="equation-8" width="256" src="https://user-images.githubusercontent.com/49260794/123195481-d2d3ff80-d4da-11eb-9554-fe972f18de8d.PNG">
> 
> In the code: <img alt="equation-9" width="774" src="https://user-images.githubusercontent.com/49260794/123195826-67d6f880-d4db-11eb-80af-cdfb5ba187df.PNG">
> 
> Why in the computation of logq, " **-** logdet_tot_q" not " **+** logdet_tot_q" ?

hi @Liujingxiu23 I understand why there is not "+" there. Do you want me to explain it?

**Comment by Subuday** [comment]
Hi @p0p4k , yeah it would be nice if you explain it. Honestly, I don't clearly understand in general the formula under log and nll.

**Comment by Liujingxiu23** [comment]
@p0p4k though long time passed, would you please explain it？ Thanks a lot

**Comment by BorisovMaksim** [comment]
@Liujingxiu23 @Subuday I think this minus is due to change of variables in normalizing flow
## General case

Let $y = f_{\theta}(z)$, where  $f_{\theta}$ is bijective and differentiable. 
 
We assume that $p_{\theta}(y| c) \sim  N (y; \mu_{\theta}(c), \sigma_{\theta}(c))$ 

Then, by rule of change-of-variables:
$$p_{\theta}(z|c) =  p_{\theta}(y| c) \left| det \frac{\partial y}{\partial z} \right| = N(y| \mu_{\theta}(c), \sigma_{\theta}(c) ) \left| det \frac{\partial y}{\partial z} \right|$$ 

Typically, $f_{\theta}(z) = f_{k} \circ f_{k-1} \circ ... \circ  f_{1}$ where $f_{i}$ are bijective and differentiable. In this model we have $k=4$ flows 

$$log\left(p_{\theta}(z|c)\right) =  log\left(p_{\theta}(y| c)\right) + \sum_{i=1}^{k} log\left| det \frac{\partial y_{i}}{\partial z_{i}} \right|$$

# Current case
Since our flow transforms noise $e_q$ to [u, v], we have

$$log\left(q_{\theta}(e_q | d, c)\right) =  log\left(q_{\theta}(u, v| d, c)\right) + \sum_{i=1}^{k} log\left| det \frac{\partial uv_{i}}{\partial {e_{q}}_{i}} \right|$$

$$log\left(q_{\theta}(u, v| d, c)\right)  = log\left(q_{\theta}(e_q | d, c)\right)  -  \sum_{i=1}^{k} log\left| det \frac{\partial uv_{i}}{\partial {e_{q}}_{i}} \right| $$
